sixtyfps_runtime/corelib/graphics/image.rs

```diff
@@ -0,0 +1,69 @@
+/* LICENSE BEGIN
+    This file is part of the SixtyFPS Project -- https://sixtyfps.io
+    Copyright (c) 2020 Olivier Goffart <olivier.goffart@sixtyfps.io>
+    Copyright (c) 2020 Simon Hausmann <simon.hausmann@sixtyfps.io>
+
+    SPDX-License-Identifier: GPL-3.0-only
+    This file is also available under commercial licensing terms.
+    Please contact info@sixtyfps.io for more information.
+LICENSE END */
+
+use crate::slice::Slice;
+use crate::{SharedString, SharedVector};
+
+/// A resource is a reference to binary data, for example images. They can be accessible on the file
+/// system or embedded in the resulting binary. Or they might be URLs to a web server and a downloaded
+/// is necessary before they can be used.
+#[derive(Clone, PartialEq, Debug)]
+#[repr(u8)]
+pub enum ImageReference {
+    /// A resource that does not represent any data.
+    None,
+    /// A resource that points to a file in the file system
+    AbsoluteFilePath(SharedString),
+    /// A resource that is embedded in the program and accessible via pointer
+    /// The format is the same as in a file
+    EmbeddedData(Slice<'static, u8>),
+    /// Raw ARGB
+    #[allow(missing_docs)]
+    EmbeddedRgbaImage { width: u32, height: u32, data: SharedVector<u32> },
+}
+
+impl Default for ImageReference {
+    fn default() -> Self {
+        ImageReference::None
+    }
+}
+
+/// Error generated if an image cannot be loaded for any reasons.
+#[derive(Default, Debug, PartialEq)]
+pub struct LoadImageError(());
+
+/// An image type that can be displayed by the Image element
+#[repr(transparent)]
+#[derive(Default, Clone, Debug, PartialEq)]
+// FIXME: the inner should be private
```

Comment:
Ah, that's public because of the interpreter/nodejs that uses `.0`?

Perhaps this could be solved using a getter on Image, like

```rust
    pub fn as_path(&self) -> Option<SharedString>
```

Reply:
Yes, still a lot of code uses the .0
Also to create the raw data which we probably do not want public API for

Reply:
As long as those are unambiguous from/into makes sense. I kind of like a function because it increases the readability of the code. Instead of "converting" an Image to a SharedString (what is that?), calling `as_path()` makes it clearer that a path is extracted (regardless of the string as return type).